### PR TITLE
FINERACT-1971: RepaymentOverDueBusinessEvent should not be sent when …

### PR DESCRIPTION
**Pull request by alberto-art3ch**
…balance is 0

## Description

Approved (orphaned) accounts with no disbursement are triggering `RepaymentOverdueDueBusinessEvent` reminder event after their maturity date. 

[FINERACT-1971](https://issues.apache.org/jira/browse/FINERACT-1971)

Screenshot 2024-04-12 at 7 22 04 a m

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/InitiateExternalAssetOwnerTransferTest.java [hunk 1]
```diff
@@ -933,6 +936,35 @@ public void saleExceptionHandling() {
         }
     }
 
+    @Test
+    public void buybackOnATobeDeclinedPendingSale() {
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/InitiateExternalAssetOwnerTransferTest.java]: Please remove this test case as it is part of an another PR.

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java [hunk 2]
```diff
@@ -71,4 +77,12 @@ public String getEnumStyledName() {
     public String getHumanReadableName() {
         return "Check loan repayment overdue";
     }
+
+    private static boolean isOverDueEventNeededToBeSent(Loan loan, Long numberOfDaysBeforeDueDateToRaiseEvent, LocalDate currentDate,
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java:84]: Please do loan status and total outstanding check as the very first step of this business step!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly review my comments!

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java [hunk 3]
```diff
@@ -49,10 +52,13 @@ public Loan execute(Loan loan) {
         }
         final LocalDate currentDate = DateUtils.getBusinessLocalDate();
         final List<LoanRepaymentScheduleInstallment> loanRepaymentScheduleInstallments = loan.getRepaymentScheduleInstallments();
+        List<LoanStatus> nonDisbursedStatuses = Arrays.asList(LoanStatus.INVALID, LoanStatus.SUBMITTED_AND_PENDING_APPROVAL,
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java:55]: Please add OVERPAID and CLOSED statuses as well.

File: fineract-provider/src/test/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStepTest.java [hunk 4]
```diff
@@ -175,4 +184,27 @@ public void givenLoanWithInstallmentOverdueAfterConfiguredDaysInLoanProductWhenS
         assertEquals(repaymentInstallment, loanPayloadForEvent);
         assertEquals(processedLoan, loanForProcessing);
     }
+
+    @Test
+    public void givenNonDisbursedLoanWhenStepExecutionThenNoBusinessEventIsRaised() {
```
Comment (adamsaghy) [inline fineract-provider/src/test/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStepTest.java]: I am not fully understand this test case, but please be sure to cover the following:
GIVEN Loan is SUBMITTED or APPROVED and outstanding amount is greater than ZERO and one of the instalment is OVERDUE
WHEN job got executed
THEN no overdue event

GIVEN Loan is ACTIVE, outstanding amount is ZERO, one of the installment is due date is matching with the actual date
WHEN job got executed
THEN no overdue event triggered

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java:84]: Done! removed the Loan status

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/InitiateExternalAssetOwnerTransferTest.java (hunk 1 above)
Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/InitiateExternalAssetOwnerTransferTest.java]: Done! Test case removed from this class

File: fineract-provider/src/test/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStepTest.java (hunk 4 above)
Comment (alberto-art3ch) [reply fineract-provider/src/test/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStepTest.java]: Done! Test cases created:

`givenActiveLoanWithZeroOutstandingWhenStepExecutionThenNoBusinessEventIsRaised`
`givenActiveLoanWithNonZeroOutstandingWhenStepExecutionThenBusinessEventIsRaised`

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java (hunk 2 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java:84]: I dont see anywhere the loan status check... and the total outstanding check was also not moved.

I didnt say **remove** them, i said move those checks to be the very first conditions of this business step!

File: fineract-provider/src/test/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStepTest.java [hunk 5]
```diff
@@ -175,4 +182,53 @@ public void givenLoanWithInstallmentOverdueAfterConfiguredDaysInLoanProductWhenS
...
+        when(loanForProcessing.getRepaymentScheduleInstallments()).thenReturn(loanRepaymentScheduleInstallments);
+        // when
+        Loan processedLoan = underTest.execute(loanForProcessing);
+        // then - No Business Event raised
+        verify(businessEventNotifierService, times(0)).notifyPostBusinessEvent(any());
+        assertEquals(processedLoan, loanForProcessing);
+    }
+
+    @Test
+    public void givenActiveLoanWithNonZeroOutstandingWhenStepExecutionThenBusinessEventIsRaised() {
```
Comment (adamsaghy) [inline fineract-provider/src/test/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStepTest.java:229]: I dont see test case to check whether the event got not raised if the loan status is SUBMITTED or APPROVED

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly see my comments!

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java:84]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java [hunk 6]
```diff
@@ -71,4 +79,12 @@ public String getEnumStyledName() {
     public String getHumanReadableName() {
         return "Check loan repayment overdue";
     }
+
+    private static boolean isOverDueEventNeededToBeSent(Loan loan, Long numberOfDaysBeforeDueDateToRaiseEvent, LocalDate currentDate,
+            LoanRepaymentScheduleInstallment repaymentScheduleInstallment, LocalDate repaymentDate) {
+        return repaymentDate.plusDays(numberOfDaysBeforeDueDateToRaiseEvent).equals(currentDate)
+                && loan.getLoanSummary().getTotalOutstanding().compareTo(BigDecimal.ZERO) > 0
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java]: Would you mind to move the `loan.getLoanSummary().getTotalOutstanding().compareTo(BigDecimal.ZERO) > 0` check to the top, next to the loan status check?

File: fineract-provider/src/test/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStepTest.java (hunk 5 above)
Comment (alberto-art3ch) [reply fineract-provider/src/test/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStepTest.java:229]: Done! Unit Test created:

`givenSubmittedLoanWhenStepExecutionThenNoBusinessEventIsRaised`
`givenApprovedLoanWhenStepExecutionThenNoBusinessEventIsRaised`

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java (hunk 6 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/cob/loan/CheckLoanRepaymentOverdueBusinessStep.java]: Done!

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**